File: DS Problems/Heap/basic.cpp
#include<bits/stdc++.h>
using namespace std;

// k-th smallest element
// k-th largest element

class Solution{

    vector<int> v = {7, 10, 4, 3, 20, 15};
    int n = v.size();
    int k = 2;

    public:

    /*
    1. Use max-heap to find smallest element
    2. Use min-heap to find largest element
    */

    void findKthSmallest(){
        priority_queue<int> pq; // max-heap
        for(int i=0;i<n;i++){
            pq.push(v[i]);
            if(pq.size()>k) pq.pop();
        }
        cout<<"K th smallest element: "<<pq.top()<<endl;
    }

    void findKthLargest(){
        priority_queue<int, vector<int>, greater<int>> pq; // max-heap
        for(int i=0;i<n;i++){
            pq.push(v[i]);
            if(pq.size()>k) pq.pop();
        }
        cout<<"K th largest element: "<<pq.top()<<endl;
    }
};

int main(){
    Solution s;
    s.findKthSmallest();
    s.findKthLargest();
}

File: DS_Problems/Heap/Basic.java
package DS_Problems.Heap;

import java.util.Collections;
import java.util.PriorityQueue;

// Find k-th smallest and k-th largest elements

public class Basic {
    public static void main(String[] args) {
        int nums[] = { 7, 10, 4, 3, 20, 15 };
        solve(nums, 3);
    }

    public static void solve(int nums[], int k) {
        PriorityQueue<Integer> min = new PriorityQueue<>();
        PriorityQueue<Integer> max = new PriorityQueue<>(Collections.reverseOrder());

        for (int i = 0; i < nums.length; i++) {
            min.add(nums[i]);
            max.add(nums[i]);

            if (min.size() > k) {
                min.poll();
            }
            if (max.size() > k) {
                max.poll();
            }
        }

        System.out.println(max.peek());
        System.out.println(min.peek());
    }
}
